compiler/rustc_ast/src/tokenstream.rs

```diff
@@ -188,6 +195,162 @@ impl<CTX> HashStable<CTX> for LazyTokenStream {
+            Some(&data.attrs)
+        } else {
+            None
+        }
+    }
+
+    pub fn to_tokenstream(self) -> TokenStream {
+        let trees: Vec<_> = self
+            .0
+            .iter()
+            .flat_map(|tree| match &tree.0 {
+                PreexpTokenTree::Token(inner) => {
+                    smallvec![(TokenTree::Token(inner.clone()), tree.1)].into_iter()
+                }
+                PreexpTokenTree::Delimited(span, delim, stream) => smallvec![(
+                    TokenTree::Delimited(
+                        *span,
+                        *delim,
+                        stream.create_token_stream().to_tokenstream()
+                    ),
+                    tree.1,
+                )]
+                .into_iter(),
+                PreexpTokenTree::Attributes(data) => {
+                    let mut outer_attrs = Vec::new();
+                    let mut inner_attrs = Vec::new();
+                    let attrs: Vec<_> = data.attrs.clone().into();
+                    for attr in attrs {
+                        match attr.style {
+                            crate::AttrStyle::Outer => {
+                                assert!(
+                                    inner_attrs.len() == 0,
+                                    "Found outer attribute {:?} after inner attrs {:?}",
+                                    attr,
+                                    inner_attrs
+                                );
+                                outer_attrs.push(attr);
+                            }
+                            crate::AttrStyle::Inner => {
+                                inner_attrs.push(attr);
+                            }
+                        }
+                    }
+
+                    let mut inner_attrs = Some(inner_attrs);
+
+                    let target_tokens = data.tokens.create_token_stream().to_tokenstream();
+                    let mut seen_braces = false;
+                    let mut in_angle = 0;
+                    let with_inner = target_tokens.0.iter().map(|(tree, spacing)| match tree {
+                        TokenTree::Token(token) => {
+                            let broken = token.kind.break_two_token_op();
+                            let token = broken.as_ref().map(|b| &b.1).unwrap_or(&token.kind);
+                            if token == &TokenKind::Lt {
+                                in_angle += 1;
+                            } else if token == &TokenKind::Gt {
+                                in_angle -= 1;
+                            }
+                            (tree.clone(), *spacing)
+                        }
+                        // FIXME - manually account for '<' and '> tokens, and make
+                        // sure that the brace group doesn't appear insde of them
+                        TokenTree::Delimited(span, delim, delim_tokens) => {
+                            if delim == &DelimToken::Brace && !seen_braces && in_angle == 0 {
```

Comment:
The `in_angle` check is wrong - it will be broken by code like:

```rust
struct Foo<const V: bool>;
trait Bar<const V: bool> {}

impl Bar<{0 < 1}> for Foo<{true}> {}
```

I need to come up with a better way of handling this.

Reply:
Perhaps using the *last* (rightmost) outermost brace-delimited group would work.